tests/bsim/bluetooth/host/att/timeout/main.c

```diff
@@ -0,0 +1,275 @@
+	bt_addr_le_t adva;
+	struct bt_conn *conn = NULL;
+	int err;
+	uint16_t handle = 0;
+
+	/* Two iterations:
+	 * 1. ATT timeout.
+	 * 2. Successful read.
+	 */
+	for (int i = 0; i < 2; i++) {
+		if (peripheral) {
+			EXPECT_ZERO(bt_testlib_adv_conn(&conn, BT_ID_DEFAULT, bt_get_name()));
+		}
+
+		if (central) {
+			EXPECT_ZERO(bt_testlib_scan_find_name(&adva, "peripheral"));
+			EXPECT_ZERO(bt_testlib_connect(&adva, &conn));
+
+			/* Establish EATT bearers. */
+			EXPECT_ZERO(bt_testlib_secure(conn, BT_SECURITY_L2));
+
+			while (bt_eatt_count(conn) == 0) {
+				k_msleep(100);
+			};
+		}
+
+		switch (i) {
+		case 0:
+			if (peripheral) {
+				trigger_att_timeout = true;
+			}
+
+			bs_sync_all_log("Ready to test ATT timeout");
+			break;
+		case 1:
+			if (peripheral) {
+				trigger_att_timeout = false;
+			}
+
+			bs_sync_all_log("Ready to test successful read");
+			break;
+		}
+
+		if (central) {
+			uint16_t actual_read_len;
+			uint16_t remote_read_send_len;
+
+			NET_BUF_SIMPLE_DEFINE(attr_value, sizeof(remote_read_send_len));
+
+			err = bt_testlib_att_read_by_type_sync(&attr_value, &actual_read_len,
+							       &handle, NULL, conn, bearer,
+							       UUID_2, 1, 0xffff);
+
+			switch (i) {
+			case 0:
+				__ASSERT(err == BT_ATT_ERR_UNLIKELY, "Unexpected error %d", err);
+				break;
+			case 1:
+				__ASSERT(!err, "Unexpected error %d", err);
+				__ASSERT(attr_value.len >= sizeof(remote_read_send_len),
+					 "Remote sent too little data.");
+				remote_read_send_len = net_buf_simple_pull_le16(&attr_value);
+				__ASSERT(remote_read_send_len == actual_read_len,
+					 "Length mismatch. %u %u",
+					 remote_read_send_len, actual_read_len);
+				break;
+			}
+		}
+
+		err = k_sem_take(&disconnected_sem, K_SECONDS(ATT_TIMEOUT_SEC + 2));
+
+		switch (i) {
+		case 0:
+			/* Here disconnect is triggered by the Central host due to ATT timeout. */
+			__ASSERT(!err, "Unexpected error %d", err);
+			conn = NULL;
+			break;
+		case 1:
+			/* Check that disconnect doesn't happen during time > ATT timeout. */
+			__ASSERT(err == -EAGAIN, "Unexpected error %d", err);
+			break;
+		}
+	}
+
+	if (central) {
+		/* This time disconnect from the peripheral. */
+		bt_conn_disconnect(conn, BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+	}
+
+	err = k_sem_take(&disconnected_sem, K_SECONDS(1));
+	__ASSERT_NO_MSG(!err);
+}
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	LOG_INF("Connected");
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	bool central = (get_device_nbr() == CENTRAL_DEVICE_NBR);
+	bool peripheral = (get_device_nbr() == PERIPHERAL_DEVICE_NBR);
+	uint8_t expected_reason;
+
+	LOG_INF("Disconnected: %u", reason);
+
+	if (central) {
+		expected_reason = BT_HCI_ERR_LOCALHOST_TERM_CONN;
+	}
+
+	if (peripheral) {
+		expected_reason = BT_HCI_ERR_REMOTE_USER_TERM_CONN;
+	}
+
+	__ASSERT(expected_reason == reason, "Unexpected reason %u", reason);
+	bt_conn_unref(conn);
```

Comment:
Where are the ref's being taken? Do you need to unref here?

Reply:
Reference is taken by testlib.

Reply:
So it is. I'm surprised that the testlib doesn't return the reference it took

Reply:
```suggestion
```
This reference is borrowed. You may not move it.

Testlib is indeed not supposed to leak references. Where does this happen?

Reply:
> Testlib is indeed not supposed to leak references. Where does this happen?

The testlib takes a reference in https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/common/testlib/src/adv.c#L31 for the peripheral and https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/common/testlib/src/connect.c#L65 for the central

It only calls `bt_conn_unref` if `bt_testlib_conn_wait_free` or `bt_testlib_disconnect` is called, but if neither of those are called, then it leaks the reference. 

Perhaps the `bt_conn_unref` here should be replaced with a call to `bt_testlib_conn_wait_free`?

Reply:
`bt_testlib_adv_conn` moves a to argument `*conn`. `bt_testlib_connect` does the same into `*connp`. Excuse the inconsistent naming of the parameters.

Reply:
The rule to follow is: `bt_conn *`-variables must be initialized to `NULL`. Updating them to `NULL` is only allowed with an accompanying move, e.g. into `bt_conn_unref`. A non-NULL `bt_conn *`-variable must be `NULL` at the end of its lifetime.

Reply:
`bt_testlib_connect` calls `bt_conn_le_create` which allocates reference to conn. This reference is dereferences in the disconnect callback.

Reply:
I was also incorrect in saying that `bt_testlib_conn_wait_free` would unreference it. 

But @PavelVPV is correct that if `bt_testlib_disconnect` isn't called, the reference taken by https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/common/testlib/src/adv.c#L31 or https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/common/testlib/src/connect.c#L65 is not returned by the testlib in case of uninitiated disconnects (which is the case here where the stack disconnects).